lucene/core/src/test/org/apache/lucene/search/TestCancellableCollector.java

```diff
@@ -0,0 +1,228 @@
+import org.apache.lucene.index.Term;
+import org.apache.lucene.store.Directory;
+import org.apache.lucene.util.BytesRef;
+import org.apache.lucene.util.LuceneTestCase;
+import org.apache.lucene.util.NamedThreadFactory;
+
+public class TestCancellableCollector extends LuceneTestCase {
+  Directory dir;
+  IndexReader reader;
+  IndexSearcher searcher;
+  ExecutorService executor = null;
+
+  @Override
+  public void setUp() throws Exception {
+    super.setUp();
+    dir = newDirectory();
+
+    RandomIndexWriter iw = new RandomIndexWriter(random(), dir);
+    Random random = random();
+    for (int i = 0; i < 100; i++) {
+      Document doc = new Document();
+      doc.add(newStringField("field", Integer.toString(i), Field.Store.NO));
+      doc.add(newStringField("field2", Boolean.toString(i % 2 == 0), Field.Store.NO));
+      doc.add(new SortedDocValuesField("field2", new BytesRef(Boolean.toString(i % 2 == 0))));
+      iw.addDocument(doc);
+
+      if (random.nextBoolean()) {
+        iw.commit();
+      }
+    }
+    reader = iw.getReader();
+    iw.close();
+
+    searcher = new IndexSearcher(reader);
+
+    executor =
+        new ThreadPoolExecutor(
+            4,
+            4,
+            0L,
+            TimeUnit.MILLISECONDS,
+            new LinkedBlockingQueue<Runnable>(),
+            new NamedThreadFactory("TestIndexSearcher"));
+  }
+
+  @Override
+  public void tearDown() throws Exception {
+    super.tearDown();
+    reader.close();
+    dir.close();
+
+    if (executor != null) {
+      executor.shutdown();
+    }
+
+    executor = null;
+  }
+
+  private CancellableCollector buildCancellableCollector(
+      final int numHits, boolean delayStart, boolean delayCollection) {
+    TopScoreDocCollector topScoreDocCollector = TopScoreDocCollector.create(numHits, null, 1);
+    CancellableCollector collector = new CancellableCollector(topScoreDocCollector);
+
+    DummyCancellableCollector dummyCancellableCollector =
+        new DummyCancellableCollector(collector, delayStart, delayCollection);
+
+    return dummyCancellableCollector;
+  }
+
+  private void executeSearchTest(
+      IndexSearcher searcher, Query query, CancellableCollector cancellableCollector, int numHits)
+      throws Exception {
+    TopDocs topDocs = searcher.search(query, numHits);
+
+    searcher.search(query, cancellableCollector);
+
+    CancellableCollector internalCancellableCollector =
+        (CancellableCollector) cancellableCollector.getInternalCollector();
+    TopScoreDocCollector topScoreDocCollector =
+        (TopScoreDocCollector) internalCancellableCollector.getInternalCollector();
+
+    assertEquals(topDocs.totalHits.value, topScoreDocCollector.totalHits);
+  }
+
+  private void cancelQuery(CancellableCollector cancellableCollector, final int sleepTime) {
+    executor.submit(
+        () -> {
+
+          // Wait for some time to let the query start
+          try {
+            if (sleepTime > 0) {
```

Comment:
I don't understand when we want to use this. It relies on non-deterministic scheduling in the threads for the test? If we need to ensure that a task has started, have it trigger a latch that the code calling this method is waiting on.